### PR TITLE
[CIT-646] Add child device support

### PR DESCRIPTION
**Pull request by rina23q**
## Proposed changes

Add a child device measurement support for c8y.
## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 1]
```diff
@@ -43,11 +43,32 @@ fn from_thin_edge_json_with_timestamp(
...
+    child_id: &str,
+) -> Result<String, CumulocityJsonError> {
+    let timestamp = WallClock.now();
+    let c8y_vec = from_thin_edge_json_with_child_with_timestamp(input, timestamp, child_id)?;
+    Ok(c8y_vec)
+}
+
+fn from_thin_edge_json_with_child_with_timestamp(
+    input: &str,
+    default_timestamp: DateTime<FixedOffset>,
```
Comment (albinsuresh) [inline mapper/cumulocity/c8y_translator_lib/src/json.rs:58]:
```suggestion
    timestamp: DateTime<FixedOffset>,
```

Wondering why it was called default in the `from_thin_edge_json_with_timestamp` method as well.

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 2]
```diff
@@ -219,4 +240,38 @@ mod tests {
         );
         }
     }
+
+    #[test_case(
+    "{\"temperature\": 23.0, \"pressure\": 220.0}\"",
+    "child1"
+    ;"input thin-edge json")]
```
Comment (albinsuresh) [inline mapper/cumulocity/c8y_translator_lib/src/json.rs:247]:
```suggestion
    ;"child device thin-edge json translation")]
```
To better reflect what the test is about;

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 3]
```diff
@@ -219,4 +240,38 @@ mod tests {
...
+                    "value": 220.0
+                }
+            }
+        });
+
+        assert_json_eq!(
+            serde_json::from_str::<serde_json::Value>(output.unwrap().as_str()).unwrap(),
+            expected_output
+        );
+    }
```
Comment (albinsuresh) [inline mapper/cumulocity/c8y_translator_lib/src/json.rs:276]: You need an additional test for the **JSON with timestamp** variant as well

File: mapper/tedge_mapper/src/converter.rs [hunk 4]
```diff
@@ -2,4 +2,16 @@ pub trait Converter: Send + Sync {
     type Error;
 
     fn convert(&self, input: &str) -> Result<String, Self::Error>;
+
+    fn convert_child_device_payload(
+        &self,
+        _input: &str,
+        _child_id: &str,
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/converter.rs:9]: Input parameters in a trait function prefixed with _ looks weird? Is it because the default implementation is ignoring it, which might be throwing a warning?

File: mapper/tedge_mapper/src/mapper.rs [hunk 5]
```diff
@@ -19,6 +20,10 @@ pub fn make_valid_topic_or_panic(topic_name: &str) -> Topic {
     Topic::new(topic_name).expect("This is a valid topic name")
 }
 
+pub fn make_valid_topic_filter_or_panic(filter_name: &str) -> TopicFilter {
+    TopicFilter::new(filter_name).expect("This is a valid topic filter name")
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/mapper.rs:24]:
```suggestion
    TopicFilter::new(filter_name).expect("This is an invalid topic filter name")
```

File: mapper/tedge_mapper/src/converter.rs (hunk 4 above)
Comment (rina23q) [reply mapper/tedge_mapper/src/converter.rs:9]: Exactly. Compiler warns unused variables because the default implementation doesn't use them.

File: mapper/tedge_mapper/src/mapper.rs (hunk 5 above)
Comment (rina23q) [reply mapper/tedge_mapper/src/mapper.rs:24]: Good catch. The original function's expect message should be changed as well.

File: mapper/tedge_mapper/src/mapper.rs [hunk 6]
```diff
@@ -86,26 +91,103 @@ impl Mapper {
...
+                    let error = MapperError::InvalidChildId { id: child_id };
+                    error!("Child ID error: {}", error);
                     self.client
                         .publish(mqtt_client::Message::new(
                             &self.config.errors_topic,
                             error.to_string(),
                         ))
                         .await?;
                 }
+                Some(child_id) => {
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/mapper.rs:117]: I'm wondering if this main mapper loop can be simplified to just a call to a convertor and publish the resultant message back to the MQTT broker, and this whole child device specific checks and manipulations can be moved as the responsibility of the `convertor`. That way this mapper will act just like a message router and the translation logic completely stays in the `convertor` which might make it easier for us to unit test quite a lot of this logic which might have to be integration tested otherwise.

For that, the convertor's convert logic will have to be changed to consume a `Message` and convert it to another `Message` as follows:
```
fn convert(&self, input: &Message) -> Result<Message, Self::Error>;
```
This might eliminate the need for two separate methods `convert()` and `convert_child_device_payload()` here.

File: mapper/tedge_mapper/src/mapper.rs [hunk 7]
```diff
@@ -86,26 +91,103 @@ impl Mapper {
...
+                        };
+                    }
+
+                    match self
+                        .converter
+                        .convert_child_device_payload(message.payload_str()?, child_id.as_str())
+                    {
+                        Ok(mapped) => {
+                            self.client
+                                .publish(mqtt_client::Message::new(&self.config.out_topic, mapped))
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/mapper.rs:136]: Since this `publish` and the `publish` for child device creation message in line no.126 are going to two different topics which don't guarantee any serialisation, isn't there the possibility of a race here where the measurement might reach the cloud first before the device creation message itself?

File: mapper/tedge_mapper/src/mapper.rs (hunk 6 above)
Comment (albinsuresh) [reply mapper/tedge_mapper/src/mapper.rs:117]: The signature of `convert` might even have to return a `Vec<Message>` instead of a single `Message` to handle the case of one inbound message resulting in the creation of multiple outbound messages as in the case of this child-device measurement handling:
```
fn convert(&self, input: &Message) -> Result<Vec<Message>, Self::Error>;
```

**Comment by rina23q** [comment]
I am convinced that integration tests will be duplicated to system tests, so I didn't add integration tests here. 
System tests => https://github.com/thin-edge/thin-edge.io/pull/576

File: mapper/tedge_mapper/src/mapper.rs [hunk 8]
```diff
@@ -86,36 +91,108 @@ impl Mapper {
...
                     self.client
                         .publish(mqtt_client::Message::new(
                             &self.config.errors_topic,
                             error.to_string(),
                         ))
                         .await?;
                 }
+                Some(child_id) => {
+                    info!("Child ID {:?}", child_id);
+                    dbg!(&children);
```
Comment (didier-wenzek) [inline mapper/tedge_mapper/src/mapper.rs:119]: `dbg!` to be removed before merge.

File: mapper/tedge_mapper/src/mapper.rs [hunk 9]
```diff
@@ -86,36 +91,108 @@ impl Mapper {
...
+                }
+                None => {
+                    // parent device
+                    match self.converter.convert(message.payload_str()?) {
+                        Ok(mapped) => {
+                            self.client
+                                .publish(mqtt_client::Message::new(&self.config.out_topic, mapped))
+                                .await?;
+                        }
+                        Err(error) => {
```
Comment (didier-wenzek) [inline mapper/tedge_mapper/src/mapper.rs:158]: This error handling is repeated for the 3 cases. Better to have the whole
`match get_child_id_from_topic(message.clone().topic.name) { ... }`
returning a `Result<Message, ConversionError>` and then publish the response on the appropriate topic.

File: mapper/tedge_mapper/src/mapper.rs [hunk 10]
```diff
@@ -86,36 +91,108 @@ impl Mapper {
...
-                        .publish(mqtt_client::Message::new(&self.config.out_topic, mapped))
-                        .await?;
-                }
-                Err(error) => {
-                    debug!("Mapping error: {}", error);
+            debug!("Topic {:?}", message.topic.name);
+
+            // ? Validation for child device ID?
+            match get_child_id_from_topic(message.clone().topic.name) {
+                Some(child_id) if child_id.is_empty() => {
```
Comment (didier-wenzek) [inline mapper/tedge_mapper/src/mapper.rs:107]: I might be easier to move the validation of the child id inside the `get_child_id_from_topic` function.

```
fn get_child_id_from_topic(topic: String) -> Result<Option<String>, ConversionError>
```

File: mapper/tedge_mapper/src/mapper.rs (hunk 6 above)
Comment (didier-wenzek) [reply mapper/tedge_mapper/src/mapper.rs:117]: The signature proposed by @albinsuresh has the huge benefit to be prepared to many use cases.

A mapper might return 0, 1, 2, or more messages for an input message.

**Review by didier-wenzek — COMMENTED** [review]
I think that it worth the cost to give a try of the signature proposed by @albinsuresh.

```
fn convert(&self, input: &Message) -> Result<Vec<Message>, Self::Error>;
```

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 11]
```diff
@@ -36,9 +36,29 @@ pub fn from_thin_edge_json(input: &str) -> Result<String, CumulocityJsonError> {
...
+    let mut serializer = serializer::C8yJsonSerializer::new(timestamp);
+    let () = parse_str(input, &mut serializer)?;
+    Ok(serializer.into_string()?)
+}
+
+/// Converts from thin-edge Json to c8y_json for child device
+pub fn from_thin_edge_json_with_child(
+    input: &str,
+    child_id: &str,
+) -> Result<String, CumulocityJsonError> {
```
[the remark this answers is not in the document]
Comment (rina23q) [reply mapper/cumulocity/c8y_translator_lib/src/json.rs:50]: I did refactoring for those 4 functions.

Public API (I didn't add `maybe_child_id` to Public APIs, especially I wanted to avoid a change in `from_thin_edge_json`, which is used in several places already)
- from_thin_edge_json(input: &str)
- from_thin_edge_json_with_child(input: &str,child_id: &str)

Private function
- from_thin_edge_json_with_timestamp(input: &str, timestamp: DateTime<FixedOffset>, maybe_child_id: Option<&str>)

File: mapper/tedge_mapper/src/c8y_converter.rs [hunk 12]
```diff
@@ -12,11 +13,40 @@ impl Converter for CumulocityConverter {
...
+        child_id: &str,
+    ) -> Result<String, Self::Error> {
+        let () = self.size_threshold.validate(input)?;
+        c8y_translator_lib::json::from_thin_edge_json_with_child(input, child_id)
+            .map_err(Into::into)
+    }
+
+    fn convert_child_device_creation(&self, child_id: &str) -> Option<Message> {
+        Some(Message::new(
+            &Topic::new("c8y/s/us").unwrap(),
```
[the remark this answers is not in the document]
Comment (rina23q) [reply mapper/tedge_mapper/src/c8y_converter.rs:29]: The change will be really useful in the future!

File: mapper/tedge_mapper/src/error.rs [hunk 13]
```diff
@@ -37,4 +40,7 @@ pub enum ConversionError {
 
     #[error(transparent)]
     SizeThresholdExceeded(#[from] SizeThresholdExceeded),
+
+    #[error(transparent)]
+    MqttClientError(#[from] MqttClientError),
```
[the remark this answers is not in the document]
Comment (rina23q) [reply mapper/tedge_mapper/src/error.rs:45]: It's leftover from past experiences.

P.S. Compiler doesn't warn unused transparent errors.

File: mapper/tedge_mapper/src/sm_c8y_mapper/tests.rs [hunk 14]
```diff
@@ -1,81 +1,83 @@
...
-
-#[tokio::test]
-#[serial]
-async fn mapper_publishes_software_update_request() {
-    // The test assures SM Mapper correctly receives software update request smartrest message on `c8y/s/ds`
-    // and converts it to thin-edge json message published on `tedge/commands/req/software/update`.
-    let broker = mqtt_tests::test_mqtt_broker();
-    let mut messages = broker
-        .messages_published_on("tedge/commands/req/software/update")
+#[cfg(test)]
```
Comment (rina23q) [inline mapper/tedge_mapper/src/sm_c8y_mapper/tests.rs:1]: The change of this file is just to add `#[cfg(test)]`. In terms of functionalities, this file is unrelated to child device support.

**Comment by rina23q** [comment]
I found a bug by myself... if user publishes invalid JSON, then publishes a valid JSON mesage, the mapper doesn't publish a child device creation message.

File: mapper/tedge_mapper/src/sm_c8y_mapper/tests.rs (hunk 14 above)
Comment (didier-wenzek) [reply mapper/tedge_mapper/src/sm_c8y_mapper/tests.rs:1]: This is not required as already stated in `mod.rs`.

This is even a bit heavy, since now the module name is `tests::tests` and tagged twice as a test module.

**Review by didier-wenzek — COMMENTED** [review]
Close to be approved. Can you please revert the change in `sm_c8y_mapper/tests.rs`.

**Review by didier-wenzek — APPROVED** [review]
Approved

File: mapper/cumulocity/c8y_translator_lib/src/serializer.rs [hunk 15]
```diff
@@ -352,4 +366,32 @@ mod tests {
...
     }
+
+    #[test]
+    fn serialize_timestamp_child_message() -> anyhow::Result<()> {
+        let timestamp = FixedOffset::east(5 * 3600)
+            .ymd(2021, 6, 22)
+            .and_hms_nano(17, 3, 14, 123456789);
+
+        let mut serializer = C8yJsonSerializer::new(timestamp, Some("child1"));
+        serializer.visit_timestamp(timestamp)?;
```
[the remark this answers is not in the document]
Comment (rina23q) [reply mapper/cumulocity/c8y_translator_lib/src/serializer.rs:377]: Added "temperature".

File: mapper/tedge_mapper/src/c8y_converter.rs [hunk 16]
```diff
@@ -1,26 +1,186 @@
...
+    fn get_mapper_config(&self) -> &MapperConfig {
+        &self.mapper_config
+    }
+
+    fn convert_messages(&mut self, input: &Message) -> Result<Vec<Message>, ConversionError> {
+        let () = self.size_threshold.validate(input.payload_str()?)?;
+
+        let mut vec: Vec<Message> = Vec::new();
+
+        let maybe_child_id = get_child_id_from_topic(input.clone().topic.name)?;
```
[the remark this answers is not in the document]
Comment (rina23q) [reply mapper/tedge_mapper/src/c8y_converter.rs]: Change the argument from `String` to `&str`.

File: mapper/tedge_mapper/src/c8y_converter.rs [hunk 17]
```diff
@@ -37,20 +95,90 @@ mod test {
...
+    fn convert_first_thin_edge_json_invalid_then_valid_with_child_id() {
+        let mut converter = Box::new(CumulocityConverter::new(SizeThreshold(16 * 1024)));
+        let in_topic = "tedge/measurements/child1";
+        let in_invalid_payload = r#"{"temp": invalid}"#;
+        let in_valid_payload = r#"{"temp": 1, "time": "2021-11-16T17:45:40.571760714+01:00"}"#;
+        let in_first_message = Message::new(&Topic::new_unchecked(in_topic), in_invalid_payload);
+        let in_second_message = Message::new(&Topic::new_unchecked(in_topic), in_valid_payload);
+
+        // First convert invalid Thin Edge JSON message.
+        let _ = converter.convert(&in_first_message);
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/c8y_converter.rs:155]: It would be better to validate the outcome of this `convert()` call as well that it's an `Err`.

File: mapper/tedge_mapper/src/c8y_converter.rs [hunk 18]
```diff
@@ -37,20 +95,90 @@ mod test {
...
+        // Second convert valid Thin Edge JSON message.
+        let second_out_messages = converter.convert(&in_second_message);
+
+        let expected_smart_rest_message = Message::new(
+            &Topic::new_unchecked("c8y/s/us"),
+            "101,child1,child1,thin-edge.io-child",
+        );
+        let expected_c8y_json_message = Message::new(
+            &Topic::new_unchecked("c8y/measurement/measurements/create"),
+            "{\"type\":\"ThinEdgeMeasurement\",\"externalSource\":{\"externalId\":\"child1\",\"type\":\"c8y_Serial\"},\"temp\":{\"temp\":{\"value\":1.0}},\"time\":\"2021-11-16T17:45:40.571760714+01:00\"}"
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/c8y_converter.rs:166]: Use `r#` formatting if possible.

File: mapper/tedge_mapper/src/converter.rs [hunk 19]
```diff
@@ -1,17 +1,43 @@
...
-        _child_id: &str,
-    ) -> Result<String, Self::Error> {
-        Ok("".to_string())
+    fn get_in_topic_filter(&self) -> &TopicFilter {
+        &self.get_mapper_config().in_topic_filter
     }
 
-    fn convert_child_device_creation(&self, _child_id: &str) -> Option<mqtt_client::Message> {
-        None
+    fn convert_messages(&mut self, input: &Message) -> Result<Vec<Message>, Self::Error>;
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/converter.rs:21]: Minor: That's a slightly confusing name as the `convert` function is the one that always produces messages, even when there are errors in the conversion. This looks more like a `convert_fallible` if that name makes any sense.

File: mapper/tedge_mapper/src/converter.rs [hunk 20]
```diff
@@ -1,17 +1,43 @@
...
+                    &self.get_mapper_config().errors_topic,
+                    error.to_string(),
+                )]
+            }
+        }
     }
 }
+
+pub fn make_valid_topic_or_panic(topic_name: &str) -> Topic {
+    Topic::new(topic_name).expect("Invalid topic name")
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/converter.rs:38]:
```suggestion
    Topic::new(topic_name).expect("Invalid topic name: {}", topic_name)
```

File: mapper/tedge_mapper/src/converter.rs [hunk 21]
```diff
@@ -1,17 +1,43 @@
...
+}
+
+pub fn make_valid_topic_filter_or_panic(filter_name: &str) -> TopicFilter {
+    TopicFilter::new(filter_name).expect("Invalid topic filter name")
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/converter.rs:42]:
```suggestion
    TopicFilter::new(filter_name).expect("Invalid topic filter name: {}", topic_name)
```

File: mapper/tedge_mapper/src/c8y_converter.rs [hunk 22]
```diff
@@ -1,26 +1,184 @@
...
+        dbg!(&second_out_messages);
+        assert_eq!(
+            second_out_messages,
+            vec![
+                expected_smart_rest_message,
+                expected_c8y_json_message.clone()
+            ]
+        );
+    }
+
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/c8y_converter.rs:177]: Worth adding one more test with measurements sent to to 2 different child device topics and make sure that device creation happens for both.

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 23]
```diff
@@ -242,33 +235,46 @@ mod tests {
     }
 
     #[test_case(
-    "{\"temperature\": 23.0, \"pressure\": 220.0}\"",
-    "child1"
-    ;"input thin-edge json")]
+    "{\"temperature\": 23.0}\"", "child1"
```
[the remark this answers is not in the document]
Comment (rina23q) [reply mapper/cumulocity/c8y_translator_lib/src/json.rs:238]: Oh yeah right, `r#...#` works. I'll change all of them.

File: mapper/tedge_mapper/src/c8y_converter.rs [hunk 24]
```diff
@@ -2,33 +2,91 @@ use crate::converter::*;
...
-            .map_err(Into::into)
+    fn convert_messages(&mut self, input: &Message) -> Result<Vec<Message>, ConversionError> {
+        let () = self.size_threshold.validate(input.payload_str()?)?;
+
+        let mut vec: Vec<Message> = Vec::new();
+
+        let maybe_child_id = get_child_id_from_topic(&input.topic.name)?;
+        match maybe_child_id {
+            Some(child_id) => {
+                let c8y_json_child_payload =
```
[the remark this answers is not in the document]
Comment (rina23q) [reply mapper/tedge_mapper/src/c8y_converter.rs:52]: We can't move it to prevent a bug: if user publishes invalid JSON, then publishes a valid JSON mesage, the mapper doesn't publish a child device creation message.

So, have to keep those order due to:
- We need to return an error when the input Thin Edge JSON is invalid before creating the child device creation message.
- Child device creation messages must be added to the vector before the measurement JSON message.

However, I can add a comment. Maybe people will wonder the order.

File: mapper/tedge_mapper/src/c8y_converter.rs [hunk 25]
```diff
@@ -2,33 +2,91 @@ use crate::converter::*;
...
+                    c8y_translator_lib::json::from_thin_edge_json_with_child(
+                        input.payload_str()?,
+                        child_id.as_str(),
+                    )?;
+
+                if !self.children.contains(child_id.as_str()) {
+                    self.children.insert(child_id.clone());
+                    vec.push(Message::new(
+                        &Topic::new_unchecked(SMARTREST_PUBLISH_TOPIC),
+                        format!("101,{},{},thin-edge.io-child", child_id, child_id),
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply mapper/tedge_mapper/src/c8y_converter.rs:62]: Indeed, this might be an issue. But for now, I would simply ignore it.

File: mapper/tedge_mapper/src/converter.rs (hunk 19 above)
Comment (didier-wenzek) [reply mapper/tedge_mapper/src/converter.rs:21]: A better name could be `try_convert`.

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 26]
```diff
@@ -235,46 +235,43 @@ mod tests {
     }
 
     #[test_case(
-    "{\"temperature\": 23.0}\"", "child1"
+    "child1",
+    r#"{"temperature": 23.0}"#,
+    json!({
```
Comment (albinsuresh) [inline mapper/cumulocity/c8y_translator_lib/src/json.rs:240]: Macro inside a macro. Nice :)

**Review by albinsuresh — APPROVED** [review]

File: mapper/tedge_mapper/src/converter.rs (hunk 20 above)
Comment (rina23q) [reply mapper/tedge_mapper/src/converter.rs:38]: It requires `format!()`. I would skip them as we use those functions only to known topics, never panics a kind of.
